skrub/_datetime_encoder.py

```diff
@@ -18,78 +23,473 @@
+    """
+    indices = []
+    index_to_format = {}
+
+    for col_idx, X_col in enumerate(X_split):
+        X_col = X_col[pd.notnull(X_col)]
+
+        # convert pd.TimeStamp to np.datetime64
+        if all(isinstance(val, pd.Timestamp) for val in X_col):
+            X_col = X_col.astype("datetime64")
+
+        if _is_column_datetime_parsable(X_col):
+            indices.append(col_idx)
+            # TODO: pass require_dayfirst to _guess_datetime_format
+            index_to_format[col_idx] = _guess_datetime_format(X_col)
+
+    return indices, index_to_format
+
+
+def _is_column_datetime_parsable(X_col):
+    """Check whether a 1d array can be converted into a \
+    :class:`pandas.DatetimeIndex`.
+
+    Parameters
+    ----------
+    X_col : array-like of shape ``(n_samples,)``
+
+    Returns
+    -------
+    is_dt_parsable : bool
+    """
+    # Remove columns of int, float or bool casted as object.
+    # Pandas < 2.0.0 raise a deprecation warning instead of an error.
+    with (
+        warnings.catch_warnings()
+        if not _is_pandas_format_mixed_available()
+        else nullcontext()
+    ):
+        if not _is_pandas_format_mixed_available():
+            warnings.simplefilter("ignore", category=DeprecationWarning)
+        try:
+            if np.array_equal(X_col, X_col.astype(np.float64)):
+                return False
+        except ValueError:
+            pass
+
+    np_dtypes_candidates = [np.object_, np.str_, np.datetime64]
+    is_type_datetime_compatible = any(
+        np.issubdtype(X_col.dtype, np_dtype) for np_dtype in np_dtypes_candidates
+    )
+    if is_type_datetime_compatible:
+        try:
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore", category=UserWarning)
+                # format=mixed parses entries individually,
+                # avoiding ValueError when both date and datetime formats
+                # are present.
+                # At this stage, the format itself doesn't matter.
+                _ = pd.to_datetime(X_col, format=MIXED_FORMAT)
+            return True
+        except (pd.errors.ParserError, ValueError):
+            pass
+    return False
+
+
+def _guess_datetime_format(X_col):
+    """
+    Parameters
+    ----------
+    X_col : ndarray of shape ``(n_samples,)``
+
+    require_dayfirst : bool, default False
+        Whether to return the dayfirst format when both dayfirst
+        and monthfirst are valid.
+
+    Returns
+    -------
+    format : str
+    """
+    if np.issubdtype(X_col.dtype, np.datetime64):
+        # We don't need to specify a parsing format
+        # for columns that are already of type datetime64.
+        return None
+
+    X_col = X_col.astype(np.object_)
+    vfunc = np.vectorize(guess_datetime_format)
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore", category=UserWarning)
+        # pd.unique handles None
+        month_first_formats = pd.unique(vfunc(X_col, dayfirst=False))
```

Comment:
Not sure if it could be a problem but note here we might end up with the string `'None'` in the result, eg for `_guess_datetime_format(np.asarray(["01/23/2021", ""]))` we get `array(['%m/%d/%Y', 'None'], dtype='<U8')` in `month_first_formats`